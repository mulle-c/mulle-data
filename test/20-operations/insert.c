#include <mulle-data/mulle-data.h>

#include <stdio.h>


static void  mulle_range_print( struct mulle_range a)
{
   if( ! mulle_range_is_valid( a))
      printf( "INVALID");
   else
      if( ! a.length)
         printf( "NULL");
      else
         printf( "%tu-%tu (%tu)", a.location, a.location + a.length - 1, a.length);
}


static void  test( struct mulle_range a, struct mulle_range b)
{
   struct mulle_range   result[ 2];
   unsigned int         i;
   unsigned int         n;
   char                 *sep = " ";

   mulle_range_print( a);
   printf( " <-insert- ");
   mulle_range_print( b);

   n = mulle_range_insert( a, b, result);
   printf( " : %u {", n);
   for( i = 0; i < n; i++)
   {
      printf( "%s", sep);
      mulle_range_print( result[ i]);
      sep =", ";
   }
   printf( "}\n");
}


int   main( int argc, char *argv[])
{
   test( mulle_range_make( 0, 0),   mulle_range_make( 0, 0));
   test( mulle_range_make( 0, 0),   mulle_range_make( 18, 48));
   test( mulle_range_make( 18, 48), mulle_range_make( 0, 0));

   test( mulle_range_make( 0, 9),   mulle_range_make( 2, 3));
   test( mulle_range_make( 2, 4),   mulle_range_make( 0, 2));
   test( mulle_range_make( 2, 4),   mulle_range_make( 6, 2));
   test( mulle_range_make( 2, 4),   mulle_range_make( 8, 2));
   test( mulle_range_make( 2, 4),   mulle_range_make( 3, 5));
   test( mulle_range_make( 2, 4),   mulle_range_make( 1, 2));

   test( mulle_range_make_invalid(), mulle_range_make( 18, 48));
   test( mulle_range_make( 18, 48),  mulle_range_make_invalid());
   test( mulle_range_make_invalid(), mulle_range_make_invalid());

   return( 0);
}
