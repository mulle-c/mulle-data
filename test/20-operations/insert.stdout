NULL <-insert- NULL : 0 {}
NULL <-insert- 18-65 (48) : 0 {}
18-65 (48) <-insert- NULL : 0 {}
0-8 (9) <-insert- 2-4 (3) : 2 { 0-1 (2), 5-11 (7)}
2-5 (4) <-insert- 0-1 (2) : 2 { NULL, 2-5 (4)}
2-5 (4) <-insert- 6-7 (2) : 2 { 2-5 (4), NULL}
2-5 (4) <-insert- 8-9 (2) : 0 {}
2-5 (4) <-insert- 3-7 (5) : 2 { 2-2 (1), 8-10 (3)}
2-5 (4) <-insert- 1-2 (2) : 2 { NULL, 3-6 (4)}
INVALID <-insert- 18-65 (48) : 0 {}
18-65 (48) <-insert- INVALID : 0 {}
INVALID <-insert- INVALID : 0 {}
